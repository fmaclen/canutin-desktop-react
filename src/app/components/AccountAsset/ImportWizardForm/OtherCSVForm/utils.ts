import { parse, format } from 'date-fns';

import { CanutinJsonTransactionType } from '@appTypes/canutin';
import { Account } from '@database/entities';
import { getBalanceGroupByAccountType } from '@database/helpers';
import { CanutinJsonType, UpdatedAccount } from '@appTypes/canutin';

import {
  SupportedDateFormatType,
} from './otherCsvConstants';
import { OtherCSVFormSubmit } from './index';

export const getTransactionsForOneAccount = (
  csvData: { [columnName: string]: string }[],
  descriptionColumn: string,
  dateColumn: string,
  dateFormat: SupportedDateFormatType,
  amountColumn: string,
  categoryColumn: string | null,
  categoryValues?: { [categoryColumnValue: string]: string }
) => {
  const transactions = csvData.map((rowData: { [x: string]: any }) => ({
    description: rowData[descriptionColumn],
    date: format(parse(rowData[dateColumn], dateFormat, new Date()), 'MM/dd/yyyy'),
    amount: Number(rowData[amountColumn]),
    excludeFromTotals: true,
    category:
      categoryColumn && categoryValues ? categoryValues[rowData[categoryColumn]] : 'Uncategorized',
  }));

  return transactions;
};

export const getTransactionsForAccountColumn = (
  accountColumn: string,
  accounts: { [accountColumnValue: string]: string } = {},
  csvData: { [columnName: string]: string }[],
  descriptionColumn: string,
  dateColumn: string,
  dateFormat: SupportedDateFormatType,
  amountColumn: string,
  categoryColumn: string | null,
  categoryValues?: { [categoryColumnValue: string]: string }
) => {
  const cFile: CanutinJsonType = { accounts: [] };
  Object.keys(accounts).forEach(accountColumnName => {
    const transactions: CanutinJsonTransactionType[] = [];

    csvData.forEach((rowData: { [x: string]: any }) => {
      if (rowData[accountColumn] === accountColumnName) {
        transactions.push({
          description: rowData[descriptionColumn],
          date: format(parse(rowData[dateColumn], dateFormat, new Date()), 'MM/dd/yyyy'),
          amount: Number(rowData[amountColumn]),
          excludeFromTotals: true,
          category:
            categoryColumn && categoryValues
              ? categoryValues[rowData[categoryColumn]]
              : 'Uncategorized',
        });
      }
    });

    const account = {
      name: accountColumnName,
      balanceGroup: getBalanceGroupByAccountType(accounts[accountColumnName]),
      accountType: accounts[accountColumnName],
      autoCalculate: true,
      transactions,
    };

    cFile.accounts.push(account);
  });

  return cFile;
};

export const getUpdatedTransactionsForExistingAccounts = (
  canutinAccounts: Account[] | null,
  accountColumn: string,
  csvData: { [columnName: string]: string }[],
  descriptionColumn: string,
  dateColumn: string,
  dateFormat: SupportedDateFormatType,
  amountColumn: string,
  categoryColumn: string | null,
  categoryValues?: { [categoryColumnValue: string]: string }
) => {
  const updatedAccounts: UpdatedAccount[] = [];
  csvData.forEach(rowData => {
    const updateAccount = canutinAccounts?.find(account => account.name === rowData[accountColumn]);
    console.log(canutinAccounts);
    console.log(rowData[accountColumn]);
    if (updateAccount) {
      const accountAlreadyCreated = updatedAccounts.find(
        account => account.id === updateAccount.id
      );

      if (!accountAlreadyCreated) {
        const transactions: CanutinJsonTransactionType[] = [
          {
            description: rowData[descriptionColumn],
            date: format(parse(rowData[dateColumn], dateFormat, new Date()), 'MM/dd/yyyy'),
            amount: Number(rowData[amountColumn]),
            excludeFromTotals: true,
            category:
              categoryColumn && categoryValues
                ? categoryValues[rowData[categoryColumn]]
                : 'Uncategorized',
          },
        ];
        updatedAccounts.push({ id: updateAccount.id, transactions });
      }

      if (accountAlreadyCreated) {
        accountAlreadyCreated.transactions.push({
          description: rowData[descriptionColumn],
          date: format(parse(rowData[dateColumn], dateFormat, new Date()), 'MM/dd/yyyy'),
          amount: Number(rowData[amountColumn]),
          excludeFromTotals: true,
          category:
            categoryColumn && categoryValues
              ? categoryValues[rowData[categoryColumn]]
              : 'Uncategorized',
        });
      }
    }
  });

  return updatedAccounts;
};

export const formToCantuinJsonFile = (
  formData: OtherCSVFormSubmit,
  csvData: any,
  canutinAccounts: Account[] | null
) => {
  const {
    accountColumn,
    amountColumn,
    categoryColumn,
    dateColumn,
    dateFormat,
    descriptionColumn,
    categories,
    accounts,
  } = formData;
  let canutinFile: CanutinJsonType = { accounts: [] };

  // New account or update account
  if (formData.account) {
    const { name, accountType, autoCalculate, balance, institution } = formData.account;
    const transactions = getTransactionsForOneAccount(
      csvData,
      descriptionColumn,
      dateColumn,
      dateFormat,
      amountColumn,
      categoryColumn,
      categories
    );
    canutinFile = {
      accounts: [
        {
          name,
          balanceGroup: getBalanceGroupByAccountType(accountType),
          accountType,
          institution,
          autoCalculate,
          balance,
          transactions,
        },
      ],
    };

    return { canutinFile };
  }

  // Multiple new accounts
  if (accountColumn) {
    canutinFile = getTransactionsForAccountColumn(
      accountColumn,
      accounts,
      csvData,
      descriptionColumn,
      dateColumn,
      dateFormat,
      amountColumn,
      categoryColumn,
      categories
    );
    const updatedAccounts = getUpdatedTransactionsForExistingAccounts(
      canutinAccounts,
      accountColumn,
      csvData,
      descriptionColumn,
      dateColumn,
      dateFormat,
      amountColumn,
      categoryColumn,
      categories
    );

    return { canutinFile, updatedAccounts };
  }

  return canutinFile;
};
