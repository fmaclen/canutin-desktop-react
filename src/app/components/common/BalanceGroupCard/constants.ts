export enum BalanceGroupCardTypeEnum {
  CASH,
  DEBT,
  INVESTMENTS,
  OTHER_ASSETS,
  NET_WORTH,
}

export const getBalanceGroupCardTitle = {
  [BalanceGroupCardTypeEnum.CASH]: 'Cash',
  [BalanceGroupCardTypeEnum.DEBT]: 'Debt',
  [BalanceGroupCardTypeEnum.INVESTMENTS]: 'Investments',
  [BalanceGroupCardTypeEnum.OTHER_ASSETS]: 'Other assets',
  [BalanceGroupCardTypeEnum.NET_WORTH]: 'Net worth',
};
