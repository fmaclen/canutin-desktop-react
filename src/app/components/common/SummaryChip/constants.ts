export enum SummaryChipTypeEnum {
  CASH,
  DEBT,
  INVESTMENTS,
  OTHER_ASSETS,
  NET_WORTH,
}

export const getSummaryChipTitle = {
  [SummaryChipTypeEnum.CASH]: 'Cash',
  [SummaryChipTypeEnum.DEBT]: 'Debt',
  [SummaryChipTypeEnum.INVESTMENTS]: 'Investments',
  [SummaryChipTypeEnum.OTHER_ASSETS]: 'Other assets',
  [SummaryChipTypeEnum.NET_WORTH]: 'Net worth',
};
