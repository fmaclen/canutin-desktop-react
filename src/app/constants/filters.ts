import {
  subMonths,
  subYears,
  startOfMonth,
  endOfMonth,
  startOfYear,
  endOfYear,
  endOfDay,
  eachMonthOfInterval,
  format,
} from 'date-fns';

const today = new Date();
const thisMonthFrom = startOfMonth(today);
const thisMonthTo = endOfMonth(today);
const thisYearFrom = startOfYear(today);
const thisYearTo = endOfYear(today);

export const filters = [
  {
    label: 'This month',
    dateFrom: thisMonthFrom,
    dateTo: thisMonthTo,
  },
  {
    label: 'Last month',
    dateFrom: subMonths(thisMonthFrom, 1),
    dateTo: subMonths(thisMonthTo, 1),
  },
  {
    label: 'Last 3 months',
    dateFrom: subMonths(today, 3),
    dateTo: endOfDay(today),
  },
  {
    label: 'Last 6 months',
    dateFrom: subMonths(today, 6),
    dateTo: endOfDay(today),
  },
  {
    label: 'Last 12 months',
    dateFrom: subMonths(today, 12),
    dateTo: endOfDay(today),
  },
  {
    label: 'Year to date',
    dateFrom: thisYearFrom,
    dateTo: endOfDay(today),
  },
  {
    label: 'Last year',
    dateFrom: subYears(thisYearFrom, 1),
    dateTo: subYears(thisYearTo, 1),
  },
  {
    label: 'Lifetime',
    dateFrom: subYears(today, 900),
    dateTo: endOfDay(today),
  },
];

export const filterOptions = filters.map(({ label, dateFrom, dateTo }) => ({
  value: { dateFrom, dateTo },
  label,
}));

export const budgetFilterOptions = eachMonthOfInterval({ end: today, start: subMonths(today, 12) })
  .map(date => ({
    value: { dateFrom: startOfMonth(date), dateTo: endOfMonth(date) },
    label: format(date, 'MMMM yyyy'),
  }))
  .reverse();
