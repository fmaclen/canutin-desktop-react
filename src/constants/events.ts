export const OPEN_CREATE_VAULT = 'openCreateVault';
export const OPEN_EXISTING_VAULT = 'openExistingVault';

// Database
export const DB_NEW_ASSET = 'dbNewAsset';
export const DB_NEW_ACCOUNT = 'dbNewAccount';
export const DB_NEW_ASSET_ACK = 'dbNewAssetAck';
export const DB_NEW_ACCOUNT_ACK = 'dbNewAccountAck';

export const DB_GET_ACCOUNTS = 'dbGetAccounts';
export const DB_GET_ACCOUNTS_ACK = 'dbGetAccountsAck';

export const DB_GET_ASSETS = 'dbGetAssets';
export const DB_GET_ASSETS_ACK = 'dbGetAssetsAck';

export const DB_GET_BALANCE_STATEMENTS = 'dbGetStatements';
export const DB_GET_BALANCE_STATEMENTS_ACK = 'dbGetStatementsAck';

export const DB_GET_TRANSACTIONS = 'dbGetTransactions';
export const DB_GET_TRANSACTIONS_ACK = 'dbGetTransactionsAck';

// Import Wizard
export const IMPORT_SOURCE_FILE = 'importSourceFile';
export const IMPORT_SOURCE_FILE_ACK = 'importSourceFileAck';
export const ANALYZE_SOURCE_FILE = 'analyzeSourceFile';
export const ANALYZE_SOURCE_FILE_ACK = 'analyzeSourceFileAck';
export const LOAD_FROM_CANUTIN_FILE = 'loadFromCanutinFile';
export const LOAD_FROM_CANUTIN_FILE_ACK = 'loadFromCanutinFileAck';
export const LOAD_FROM_OTHER_CSV = 'loadFromOtherCsv';
export const LOAD_FROM_OTHER_CSV_ACK = 'loadFromOtherCsvAck';
export const LOADING_CSV = 'loadingCsv';
