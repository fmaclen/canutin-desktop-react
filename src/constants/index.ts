export const DATABASE_PATH = 'databasePath';
export const DATABASE_CONNECTED = 'databaseConnected';
export const DATABASE_DOES_NOT_EXIST = 'databaseDoesNotExist';
export const DATABASE_NOT_DETECTED = 'databaseNotDetected';
export const NEW_DATABASE = 'newDatabase';

// Canutin file
export const CANUTIN_FILE_DATE_FORMAT = 'MM/dd/yyyy';
