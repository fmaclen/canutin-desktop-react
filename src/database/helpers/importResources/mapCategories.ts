// TODO: Complete this, https://github.com/Canutin/desktop/issues/49
const mapCategories = (category: string) => {
  switch (category) {
    case 'Auto & Transport':
    case 'Auto Payment':
      return 'Automotive';

    case 'Auto Insurance':
      return 'Insurance';

    case 'Gas & Fuel':
      return 'Gas stations';

    case 'Parking':
      return 'Parking';

    case 'Public Transportation':
      return 'Public Transportation';

    case 'Service & parts':
      return 'Service & parts';

    case 'Bills & Utilities':
    case 'Utilities':
      return 'Utilities';

    case 'Home Phone':
    case 'Mobile Phone':
    case 'Internet':
      return 'Internet & phone';

    case 'Television':
      return 'Television';

    case 'Business & services':
    case 'Advertising':
      return 'Business & services';

    default:
      return 'Uncategorized';
  }
};

export default mapCategories;
