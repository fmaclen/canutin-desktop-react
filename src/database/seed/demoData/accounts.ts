import { BalanceGroupEnum } from '@enums/balanceGroup.enum';

export const accountCheckingDetails = {
  name: "Alice's Checking",
  balanceGroup: BalanceGroupEnum.CASH,
  accountType: 'checking',
  autoCalculated: true,
  closed: false,
  officialName: 'Ransack High-Yield Checking',
  institution: 'Ransack Bank',
};

export const accountSavingsDetails = {
  name: 'Emergency Fund',
  balanceGroup: BalanceGroupEnum.CASH,
  accountType: 'savings',
  autoCalculated: true,
  closed: false,
  officialName: 'Ransack Savings Plus',
  institution: 'Ransack Bank',
};

export const accountCreditCardDetails = {
  name: "Bob's Juggernaut Visa",
  balanceGroup: BalanceGroupEnum.DEBT,
  accountType: 'credit card',
  autoCalculated: true,
  closed: false,
  officialName: 'Juggernaut Cash Back Rewards',
  institution: 'Juggernaut Bank',
};

export const accountAutoLoanDetails = {
  name: 'Toyota Auto Loan',
  balanceGroup: BalanceGroupEnum.DEBT,
  accountType: 'auto',
  autoCalculated: false,
  closed: false,
  institution: 'Toyota Financial Services',
};

export const accountRothIraDetails = {
  name: "Alice's Roth IRA",
  balanceGroup: BalanceGroupEnum.INVESTMENTS,
  accountType: 'roth',
  autoCalculated: false,
  closed: false,
  officialName: 'Loot Wealth Roth IRA',
  institution: 'Loot Financial',
};

export const account401kDetails = {
  name: "Bob's 401k",
  balanceGroup: BalanceGroupEnum.INVESTMENTS,
  accountType: '401k',
  autoCalculated: false,
  closed: false,
  officialName: 'Loot Wealth 401k',
  institution: 'Loot Financial',
};

export const accountWalletDetails = {
  name: 'Wallet',
  balanceGroup: BalanceGroupEnum.CASH,
  accountType: 'cash',
  autoCalculated: false,
  closed: false,
};
