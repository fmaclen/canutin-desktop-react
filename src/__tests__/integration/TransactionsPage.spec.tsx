import { screen, waitFor } from '@testing-library/react';
import { ipcRenderer } from 'electron';
import userEvent from '@testing-library/user-event';
import selectEvent from 'react-select-event';
import 'jest-styled-components';
import { endOfDay, startOfDay } from 'date-fns';

// Fixes `ReferenceError: regeneratorRuntime is not defined` error on `useAsyncDebounce`.
// REF: https://github.com/tannerlinsley/react-table/issues/2071
import 'regenerator-runtime/runtime';

import { dateInUTC } from '@app/utils/date.utils';
import { FILTER_TRANSACTIONS } from '@constants/events';
import { accountCheckingDetails } from '@database/seed/demoData/accounts';
import { accountCheckingTransactionSet } from '@database/seed/demoData/transactions';
import { filters } from '@app/constants/filters';
import mapCategories from '@database/helpers/importResources/mapCategories';
import { initAppWith } from '@tests/utils/initApp.utils';
import { seedMinimumAccount } from '@tests/factories/entitiesFactory';

describe('Transactions tests', () => {
  test("Sidebar link can't be clicked if no accounts or assets are present", async () => {
    initAppWith({});
    const transactionsSidebarLink = screen.getByTestId('sidebar-transactions');
    expect(transactionsSidebarLink).toHaveAttribute('disabled');

    userEvent.click(transactionsSidebarLink);
    expect(transactionsSidebarLink).not.toHaveAttribute('active', '1');
  });

  test('Transactions page displays an empty view when no enough data is available', async () => {
    initAppWith({ accounts: seedMinimumAccount });
    const transactionsSidebarLink = screen.getByTestId('sidebar-transactions');
    expect(transactionsSidebarLink).toHaveAttribute('active', '0');
    expect(transactionsSidebarLink).not.toHaveAttribute('disabled');
    expect(transactionsSidebarLink).toHaveAttribute('href', '#/transactions');

    userEvent.click(transactionsSidebarLink);
    expect(transactionsSidebarLink).toHaveAttribute('active', '1');

    const scrollViewTransactions = screen.getByTestId('scrollview-transactions');
    expect(scrollViewTransactions).toMatchSnapshot();
  });

  test('Transactions page displays the correct data', async () => {
    const seedTransactionsThisMonth = accountCheckingTransactionSet()
      .filter(transaction => {
        // Mimic the logic in `TransactionRepository.getFilterTransactions()`
        const dateFrom = startOfDay(dateInUTC(filters[0].dateFrom));
        const dateTo = endOfDay(dateInUTC(filters[0].dateTo));
        return transaction.date >= dateFrom && transaction.date <= dateTo;
      })
      .map(transaction => ({
        ...transaction,
        account: { ...accountCheckingDetails },
        category: { name: mapCategories(transaction.categoryName) },
      }));

    initAppWith({ accounts: seedMinimumAccount, filterTransactions: seedTransactionsThisMonth });
    userEvent.click(screen.getByTestId('sidebar-big-picture')); // Resets path back to the default
    const transactionsSidebarLink = screen.getByTestId('sidebar-transactions');
    expect(transactionsSidebarLink).toHaveAttribute('active', '0');
    expect(transactionsSidebarLink).not.toHaveAttribute('disabled');

    userEvent.click(transactionsSidebarLink);
    expect(transactionsSidebarLink).toHaveAttribute('active', '1');

    const cardTransactions = screen.getByTestId('card-transactions');
    expect(cardTransactions).toHaveTextContent('Transactions');
    expect(cardTransactions).toHaveTextContent('8');

    let cardNetBalance = screen.getByTestId('card-net-balance');
    expect(cardNetBalance).toHaveTextContent('Net balance');
    expect(cardNetBalance).toHaveTextContent('$350');
    expect(cardNetBalance).not.toHaveTextContent('-$350');

    let rowTransactions = screen.getAllByTestId('row-transaction');
    expect(rowTransactions.length).toBe(8);
    expect(rowTransactions[0]).toHaveTextContent('Toyota - TFS Payment');
    expect(rowTransactions[0]).toHaveTextContent('Automotive');
    expect(rowTransactions[0]).toHaveTextContent("Alice's Checking");
    expect(rowTransactions[0]).toHaveTextContent('-$500');
    expect(rowTransactions[7]).toHaveTextContent('Westside Apartments');
    expect(rowTransactions[7]).toHaveTextContent('Housing');
    expect(rowTransactions[7]).toHaveTextContent("Alice's Checking");
    expect(rowTransactions[7]).toHaveTextContent('-$2,250');

    const tableHeaderDate = screen.getByText('Date');
    const tableHeaderAmount = screen.getByText('Amount');
    expect(tableHeaderDate).toBeVisible();
    expect(tableHeaderAmount).toBeVisible();
    expect(screen.getByText('Description')).toBeVisible();
    expect(screen.getByText('Category')).toBeVisible();
    expect(screen.getByText('Account')).toBeVisible();

    // Sort by "Date" (ascending)
    userEvent.click(tableHeaderDate);
    rowTransactions = screen.getAllByTestId('row-transaction');
    expect(rowTransactions[0]).toHaveTextContent('Westside Apartments');
    expect(rowTransactions[0]).toHaveTextContent('-$2,250');
    expect(rowTransactions[0]).not.toHaveTextContent('Toyota - TFS Payment');

    // Sort by "Amount" (ascending)
    userEvent.click(tableHeaderAmount);
    rowTransactions = screen.getAllByTestId('row-transaction');
    expect(rowTransactions[0]).toHaveTextContent('-$2,250');
    expect(rowTransactions[0]).not.toHaveTextContent('$2,800');

    // Sort by "Amount" (descending)
    userEvent.click(tableHeaderAmount);
    rowTransactions = screen.getAllByTestId('row-transaction');
    expect(rowTransactions[0]).toHaveTextContent('$2,800');
    expect(rowTransactions[0]).not.toHaveTextContent('-$2,250');
    expect(rowTransactions[7]).toHaveTextContent('-$2,250');

    // Filter by "Credits"
    userEvent.click(screen.getByText(/Credits/i));
    cardNetBalance = screen.getByTestId('card-net-balance');
    expect(cardNetBalance).toHaveTextContent('$5,600');

    rowTransactions = screen.getAllByTestId('row-transaction');
    expect(rowTransactions.length).toBe(3);
    expect(rowTransactions[0]).toHaveTextContent('$2,800');
    expect(rowTransactions[2]).toHaveTextContent('$0');
    expect(rowTransactions[2]).not.toHaveTextContent('-$2,250');

    // Filter by "Debits"
    userEvent.click(screen.getByText(/Debits/i));
    cardNetBalance = screen.getByTestId('card-net-balance');
    expect(cardNetBalance).toHaveTextContent('-$5,250');

    rowTransactions = screen.getAllByTestId('row-transaction');
    expect(rowTransactions.length).toBe(5);
    expect(rowTransactions[0]).toHaveTextContent('-$250');
    expect(rowTransactions[0]).not.toHaveTextContent('$2,800');
    expect(rowTransactions[4]).toHaveTextContent('-$2,250');

    // Filter by "All"
    userEvent.click(screen.getByText(/All/i));
    cardNetBalance = screen.getByTestId('card-net-balance');
    expect(cardNetBalance).toHaveTextContent('$350');

    rowTransactions = screen.getAllByTestId('row-transaction');
    expect(rowTransactions.length).toBe(8);
    expect(rowTransactions[0]).toHaveTextContent('$2,800');
    expect(rowTransactions[7]).toHaveTextContent('-$2,250');

    // Filter by keyword
    const inputSearch = screen.getByPlaceholderText(
      'Type to filter by date, description, category, account or amount'
    );
    userEvent.type(inputSearch, 'transfer to');
    await waitFor(() => {
      expect(screen.getAllByTestId('row-transaction').length).toBe(3);

      rowTransactions = screen.getAllByTestId('row-transaction');
      expect(rowTransactions.length).toBe(3);
      expect(rowTransactions[0].textContent).toContain('Transfer to MegaCoin Exchange');
      expect(rowTransactions[1].textContent).toContain('Transfer to Ransack Savings');
      expect(rowTransactions[2].textContent).toContain('Transfer to Loot Financial');
    });

    // Change date filter
    const selectLast3Months = screen.getByText('Last 3 months');
    await selectEvent.openMenu(selectLast3Months);
    expect(screen.getByText('This month')).toBeVisible();
    expect(screen.getByText('Last 6 months')).toBeVisible();
    expect(screen.getByText('Last 12 months')).toBeVisible();
    expect(screen.getByText('Year to date')).toBeVisible();
    expect(screen.getByText('Last year')).toBeVisible();
    expect(screen.getByText('Lifetime')).toBeVisible();

    // FIXME: onChange event is not fired when an option in `<CustomSelect>` is selected
    // await selectEvent.select(selectLast3Months, 'Last 6 months');
    // const spyOnIpcRenderer = jest.spyOn(ipcRenderer, 'send');
    // const { dateFrom, dateTo } = filters.find(filter => {
    //   return filter.label === 'Last 6 months';
    // })!;
    // await waitFor(() => {
    //   expect(screen.queryByText('Last 3 months')).not.toBeVisible();
    //   expect(spyOnIpcRenderer).toHaveBeenCalledWith(FILTER_TRANSACTIONS, {
    //     dateFrom: dateFrom,
    //     dateTo: dateTo,
    //   });
    // });
  });

  // TODO: add test for "add transaction" flow
});
