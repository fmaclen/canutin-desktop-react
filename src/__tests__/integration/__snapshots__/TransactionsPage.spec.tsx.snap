// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions tests Transactions page displays an empty view when no enough data is available 1`] = `
<div
  class="hHSwWn"
  data-testid="scrollview-transactions"
>
  <header
    class="kVxPac"
  >
    <div
      class="cbdbMT"
    >
      <div
        class="jGiENK"
      >
        <h1
          class="wbtfl"
          title="Transactions"
        >
          Transactions
        </h1>
      </div>
      <nav>
        <nav
          class="evIijI"
        >
          <button
            class="eVxMpZ"
            type="button"
          >
            Add transaction
          </button>
          <button
            class="eVxMpZ"
            type="button"
          >
            Import
          </button>
          <hr
            class="eJgmMW"
          />
          <div
            class="fldryk iTItLc css-2b097c-container"
          >
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class="select__control css-yk16xz-control"
            >
              <div
                class="select__value-container select__value-container--has-value css-g1d714-ValueContainer"
              >
                <div
                  class="select__single-value css-1uccc91-singleValue"
                >
                  Last 3 months
                </div>
                <input
                  aria-autocomplete="list"
                  class="css-wmatm6-dummyInput-DummyInput"
                  id="react-select-2-input"
                  readonly=""
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class="select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="select__indicator-separator css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="select__indicator select__dropdown-indicator css-tlfecz-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </nav>
      </nav>
    </div>
  </header>
  <main
    class="iLArhq"
  >
    <div
      class="dOahFo"
    >
      <header
        class="hfAFDf"
      >
        <h2
          class="dYdCgw"
        >
          Browse transactions
        </h2>
        <div>
          <nav
            class="dDdRjc"
          >
            <div
              class="dAAFPm"
            >
              all
            </div>
            <div
              class="gDxUWx"
            >
              credits
            </div>
            <div
              class="gDxUWx"
            >
              debits
            </div>
          </nav>
        </div>
      </header>
      <div
        class="ftEOHx"
      >
        <nav
          class="bvkVPF"
        >
          <input
            class="hvhRRC"
            placeholder="Type to filter by date, description, category, account or amount"
            value=""
          />
          <div
            class="eRlbCM"
          >
            <div
              class="hzBHfb"
              data-testid="card-transactions"
            >
              <p
                class="bXnesS"
              >
                Transactions
              </p>
              <p
                class="hfeHrb"
              >
                0
              </p>
            </div>
            <div
              class="hzBHfb"
              data-testid="card-net-balance"
            >
              <p
                class="bXnesS"
              >
                Net balance
              </p>
              <span
                appearance="0"
                class="lcpzDX"
              >
                $0
              </span>
            </div>
          </div>
        </nav>
        <table
          class="epmggY"
          role="table"
        >
          <thead>
            <tr
              class="dlazae"
              role="row"
            >
              <th
                class="fjGJkg"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div>
                  Date
                  <div
                    class="bYMPbw"
                  >
                    â–¼
                  </div>
                </div>
              </th>
              <th
                class="fCiLLn"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div>
                  Description
                  <div
                    class="bYMPbw"
                  />
                </div>
              </th>
              <th
                class="fCiLLn"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div>
                  Category
                  <div
                    class="bYMPbw"
                  />
                </div>
              </th>
              <th
                class="fCiLLn"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div>
                  Account
                  <div
                    class="bYMPbw"
                  />
                </div>
              </th>
              <th
                class="glSWPn"
                colspan="1"
                role="columnheader"
                style="cursor: pointer;"
                title="Toggle SortBy"
              >
                <div>
                  Amount
                  <div
                    class="bYMPbw"
                  />
                </div>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="ggkxz"
                colspan="1000"
              >
                <div
                  class="dKuPRQ fCXMoP"
                >
                  No transactions were found
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </main>
</div>
`;
